### ISSUE TITLE
QueryTimeout Not Cancelling Query

### ISSUE THREAD

**Issue by kevinamasur** [microsoft/mssql-jdbc#1607]
## Driver version
9.2.1.jre8

## SQL Server version
Microsoft SQL Server 2017 (RTM-CU13) (KB4466404) - 14.0.3048.4 (X64) 

## JAVA/JVM version
openjdk version "1.8.0_232"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_232-b09)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.232-b09, mixed mode)

## Database Connection Pool
Apache DBCP 2.7.0

## Problem Description
In our system we set the socketTimout on the jdbc connection url.  However, for individual database calls we also set the queryTimeout on the individual statement.  Generally, the queryTimeout is set to a lower timeout value than the socketTimeout.

We noticed that some long running queries in our system are left running long after the query timeout is hit.  However, if the socketTimeout is hit instead, the connection is killed, which in turn cancels the query.

This appears to be because when the queryTimeout is hit, the connection is not closed.  However, when the socketTimeout is hit, the connection is closed.  So because the query isn't cancelled on the database, and the connection isn't closed by the driver, the connection is put back into our connection pool and the query is left running against our database.

I was able to make an example test for each case:

Query Timeout:
```
		Connection conn = connectionManager.getConnection();
		PreparedStatement preparedStatement = conn.prepareStatement("WAITFOR DELAY '0:01:00'");
		preparedStatement.setQueryTimeout(1);
		try {
			preparedStatement.execute();
			fail();
		} catch (SQLException e) {
			logger.error("Query Timeout", e);
		}
		assertTrue(conn.isClosed()); //Fails, pausing here and checking the database shows query is still executing
```

Socket / Network Timeout:
```
		Connection conn = connectionManager.getConnection();
		conn.setNetworkTimeout(null, 1000);
		try {
			conn.prepareStatement("WAITFOR DELAY '0:01:00'").execute();
			fail();
		} catch (SQLException e) {
			logger.error("Network Timeout", e);
		}
		assertTrue(conn.isClosed()); // Passes
```

Is this the expected or intended behavior?  

If this is the intended behavior, why would you use the query timeout rather than the socket/network timeout if the query is left running on the database?

This is an older system, it is possible something about our system configuration or setup is causing this issue instead.

**Comment by lilgreenbird** [comment]
hi @kevinamasur 

`queryTimeout` is the number of seconds to wait before a timeout will occur on a query, it is the time before which the command must complete before the driver will report a timeout, it does not actually cancel the query execution. `setNetworkTimeout` sets the actual TCP socket timeout (SO_TIMEOUT) value so the socket read will actually timeout and you will get a java.net.SocketTimeoutException on the socket and the connection will no longer be valid.

**Comment by lilgreenbird** [comment]
`queryTimeout` is the number of seconds to wait before a timeout will occur on a query, it is the time before which the command must complete before it's interrupted. `cancelQueryTimeout` is the time to cancel the query. This property was introduced to fix the issue #525 where the `querytimeout` hangs and does not throw an exception due to silent dropping of the connection. If this is set, the driver will wait for a total amount of `queryTimeout + cancelQueryTimeout` then terminate the connection. Note that when using this you will need to set both these properties as `cancelQueryTimeout` is only applicable if `queryTimeout` is also set. These timeouts are at the application level managed by the driver. 

As mentioned previously, `setNetworkTimeout` sets the actual TCP socket timeout (SO_TIMEOUT) value so when that timeout is exceeded the socket read will actually timeout and you will get a java.net.SocketTimeoutException. Since a TCP socket is an endpoint of a connection, when a socket times out, that connection will obviously no longer be valid.

**Comment by lilgreenbird** [comment]
closing due to inactivity. Please feel free to request re-opening if you have more questions.
